slides/E-async-web.md

```diff
@@ -0,0 +1,892 @@
+
+Today, we're focusing on concurrency: _asynchronous programming_
+
+---
+layout: default
+---
+
+# What's async?
+
+- Concurrent programming model
+- Very suitable for running large number of I/O bound tasks
+  - like web servers!
+- Look and feel* of synchronous code through `async`/`await` syntax
+
+**Well, not perfectly. We'll go into that*
+
+---
+layout: default
+---
+
+# Async vs OS Threads
+
+|                      | <span style="color: red">**Async**</span> | <span style="color: blue">**OS Threads**</span> |
+| -------------------- | ----------------------------------------- | ----------------------------------------------- |
+| Spawning & switching | Cheap                                     | Expensive                                       |
+| Blocking is ok       | No                                        | Yes                                             |
+| Usage                | I/O bound tasks (web servers)             | CPU-bound tasks (Number crunching)              |
+| Reuse sync code      | No                                        | Yes                                             |
+
+[What Color is Your Function? ](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/)
+
+---
+layout: default
+---
+
+# What `async` looks like in Rust
+To get an idea
+
+```rust
+/// An async function
+async fn run() -> anyhow::Result<()> {
+    /// Await loading and parsing config file
+    let config = load_config(CONFIG_PATH).await?;
+    /// Await scraping job
+    let data = scrape(&config.urls).await?;
+    data.report();
+    Ok(())
+}
+
+fn main() {
+    // Set up a `tokio` runtime with default configurations
+    let runtime = tokio::runtime::Runtime::new().unwrap();
+    // Run a Future to completion
+    runtime.block_on(run());
+}
+```
+
+*Question: What stands out to you? What strikes you as odd?*
+
+---
+layout: default
+---
+
+# Async in Rust
+
+- Revolve around `Future` trait (~like JS `Promise`, C# `Task`)  
+  &rarr; `async fn`s return `Future`s
+
+- `Future`s are inert
+- `async` is zero-cost
+- No built-in runtime
+- Single- or multithreaded execution
+- Can be mixed with other concurrency models
+- Relatively new and lacks some features and nice diagnostics
+
+---
+layout: default
+---
+
+# State of the `async` art
+What you can expect doing `async` Rust
+
+- Blazingly fast applications
+- More interaction with advanced language features
+- Compatibility issues (re: colored functions)
+- Faster evolving ecosystem
+- `async fn` in Traits not stable
+
+*But still a work in progress*
+
+---
+layout: default
+---
+
+# Support of `async`
+
+- Fundamental types and traits are in `std`
+- `async`/`await` are native to the language
+- Utilities/extensions in `futures` crate
+- Async runtimes are third party
```

Comment:
From an outsiders perspective, that is the hardest part to understand. Async is in Rust, but the runtimes are third party? Quoi?